### PR TITLE
Wrap `libnfs4` directly from python instead of having C code in-between

### PR DESCRIPTION
**Pull request by Chenlibo**
1. Move all the logic in `test/pysfs.c` to `test/sfs.py`. 
2. Remove `test/pysfs.c` and relevant Makefile lines
3. Open and read existing files work as expected. 
4. Add `sfs_test.py` to test read-only flag `r` and write-only flag `w`.

### CODE REVIEW

File: test/sfs.py [hunk 1]
```diff
@@ -1,28 +1,87 @@
...
+    p.mode = 0o666
+    
+    flags = 0
+    for c in mode:
+        if c == 'r':
+            flags |= NFS4_RDONLY
+        elif c == 'w':
+            flags |= (NFS4_TRUNC | NFS4_WRONLY | NFS4_CREAT)
+        elif c == 'a':
+            flags |= (NFS4_WRONLY | NFS4_CREAT)
```
Comment (jssmith) [inline test/sfs.py:71]: let's add an else error condition, raise `NotImplementedError` if we're still working on it.

File: test/sfs.py [hunk 2]
```diff
@@ -1,28 +1,87 @@
...
+    
 def open(file_name, mode='r'):
-    fp = c_helper.open_file_py(file_name.encode('utf-8'), mode.encode('utf-8'))
-    return FileObjectWrapper(fp)
-
-def read(size):
-    c_helper.read_file_py(size)
+    #TODO: needs to validate MODE
+    p = Nfs4_properties_struct()
+    p.mask = 1<<33
```
Comment (jssmith) [inline test/sfs.py]: Can we document mask / say what it means?

File: test/sfs.py [hunk 3]
```diff
@@ -43,11 +102,20 @@ def seek(self, pos, whence=0):
...
     def read(self, size):
-        res = c_helper.read_file_py(self._fp, self._pos, size)
+        buffer = ctypes.create_string_buffer(size)
+        read_status = c_helper.nfs4_pread(self._file, buffer, self._pos, size)
+        if read_status != NFS4_OK:
+            print("Failed to read file: " + c_helper.nfs4_error_string(client))
+            return
         self._pos += size
-        return res
+        return bytearray(buffer)
```
Comment (jssmith) [inline test/sfs.py]: Does this create a copy? Would like to avoid copies if possible.

**Review by jssmith — COMMENTED** [review]
Overall this looks quite good. Please address the questions / comments.

File: test/sfs.py (hunk 1 above)
Comment (Chenlibo) [reply test/sfs.py:71]: resolved

File: test/sfs.py (hunk 2 above)
Comment (Chenlibo) [reply test/sfs.py]: addressed by replacing it with `NFS4_PROP_MODE`

File: test/sfs_test.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        #'r' open for reading (default)
+        # can read, unable to write, error on file does not exist, 
+        # error on missing read permission (including directory tree)
+        
+        # 1. generate a 15-character-long string as filename, chance that it exists is close to zero.
+        # Open a file with such filename should raise a FileNotFoundError exception
+        filename = '/'+''.join(random.choices(string.ascii_uppercase + string.digits, k=15))
+        self.assertRaises(FileNotFoundError, sfs.open, filename, "r") # error on non-exist file
+        
+        # 2. check if can read an existing file, make sure you have file 'haha.txt' on your efs root directory
```
Comment (scottnuma) [inline test/sfs_test.py]: it looks like haha.txt became test.txt?

**Review by scottnuma — APPROVED** [review]
Looks solid

File: test/sfs_test.py (hunk 4 above)
Comment (Chenlibo) [reply test/sfs_test.py]: ahhhh, I fogot to update the comment. Good catch!

Comment (Chenlibo) [reply test/sfs_test.py]: addressed

File: test/sfs.py (hunk 3 above)
Comment (Chenlibo) [reply test/sfs.py]: addressed!

File: test/sfs.py [hunk 5]
```diff
@@ -79,7 +79,7 @@ def open(file_name, mode='r'):
             raise FileNotFoundError("[Errno 2] No such file or directory: " + "'" + file_name + "'")
         if error_code == NFS4_EACCES:
             raise PermissionError("[Errno 13] Permission denied: " + "'" + file_name + "'")
-        print("Failed to open " + file_name + ": " + c_helper.nfs4_error_string(client))
+        print("Failed to open " + file_name + ": " + c_helper.nfs4_error_string(client).decode(encoding='utf-8', errors='ignore'))
```
Comment (jssmith) [inline test/sfs.py]: I'm curious, what errors are we ignoring here?

**Review by jssmith — COMMENTED** [review]
Let's not add `errors='ignore'` on utf-8 decoding. Invalid utf-8 is coming from the underlying library and that should raise an error here rather than silently discarding it.

File: test/sfs.py (hunk 5 above)
Comment (Chenlibo) [reply test/sfs.py]: non-utf-8 bytes

**Pull request merged**